tests/qasm3/test_loop.py

```diff
@@ -332,3 +332,134 @@ def test_convert_qasm3_for_loop_unsupported_type(caplog):
+    bit[2] c;
+    int i = 0;
+    while (i < 2) {
+        h q[i];
+        if (i == 0) {
+            i += 1;
+            continue;
+        }
+        if (i == 1) {
+            break;
+        }
+        i += 1;
+    }
+    measure q -> c;
+    """
+    result = loads(qasm_str)
+    result.unroll()
+    assert result.num_qubits == 2
+    assert result.num_clbits == 2
+    check_single_qubit_gate_op(result.unrolled_ast, 2, [0, 1], "h")
+
+
+def test_while_loop_limit_exceeded():
+    """Test that exceeding the loop limit raises LoopLimitExceeded."""
+    qasm_str = """
+    OPENQASM 3.0;
+    qubit q;
+    int i = 0;
+    while (i < 10) {
+        i += 1;
+    }
+    """
+    result = loads(qasm_str)
+    # Set loop_limit to 5 to force exception
+    with pytest.raises(LoopLimitExceeded):
+        result.unroll(loop_limit=5)
+
+
+def test_while_loop_unroll_qasm_output():
+    """Test that unrolling a while loop produces the expected QASM string."""
+    qasm_str = """
+    OPENQASM 3.0;
+    qubit q;
+    int i = 0;
+    while (i < 2) {
+        h q;
+        i += 1;
+    }
+    """
+    expected = [
+        'h q;',
+        'i += 1;',
+        'h q;',
+        'i += 1;'
+    ]
+    result = loads(qasm_str)
+    result.unroll()
+    # Check that the unrolled AST contains the expected sequence
+    stmts = [str(s) for s in result.unrolled_ast if hasattr(s, '__str__')]
+    for e in expected:
+        assert any(e in s for s in stmts)
+
+
+def test_empty_while_loop_ignored():
+    """Test that an empty while loop is ignored (no effect)."""
+    qasm_str = """
+    OPENQASM 3.0;
+    qubit q;
+    int i = 0;
+    while (i < 0) {
+    }
+    h q;
+    """
+    result = loads(qasm_str)
+    result.unroll()
+    stmts = [str(s) for s in result.unrolled_ast if hasattr(s, '__str__')]
+    assert any('h q' in s for s in stmts)
+    # No extra statements from the while loop
+    assert len(stmts) == 1
+
+
+def test_nested_while_loops_break_continue():
+    """Test nested while loops: break/continue in inner loop does not affect outer loop."""
+    qasm_str = """
+    OPENQASM 3.0;
+    qubit q;
+    int i = 0;
+    int j = 0;
+    while (i < 2) {
+        j = 0;
+        while (j < 2) {
+            if (j == 1) {
+                break;
+            }
+            j += 1;
+        }
+        i += 1;
+    }
+    h q;
+    """
+    result = loads(qasm_str)
+    result.unroll()
+    stmts = [str(s) for s in result.unrolled_ast if hasattr(s, '__str__')]
+    assert any('h q' in s for s in stmts)
+
+
+def test_mixed_for_while_loops():
+    """Test a for loop inside a while loop and vice versa."""
+    qasm_str = """
+    OPENQASM 3.0;
+    qubit[2] q;
+    int i = 0;
+    while (i < 2) {
+        for int j in [0, 1] {
+            h q[j];
+        }
+        i += 1;
+    }
+    """
+    result = loads(qasm_str)
+    result.unroll()
+    stmts = [str(s) for s in result.unrolled_ast if hasattr(s, '__str__')]
+    assert sum('h q' in s for s in stmts) == 4
```

Comment:
please validate qubit indices as well with `check_single_qubit_op` (as you have done in [above test](https://github.com/qBraid/pyqasm/pull/199/files#diff-9188f01a8792f49f21cfe6c2c54a48cb360bd1dc3baff3cf33da5f922789ecccR362)). 

Similar comment applies to all the new tests